{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exam by Álvaro Martínez Petit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1\n",
    "\n",
    "If both files contain the Locus Code in the same order, we could compare the Locus Code of a line from the first file to the Locus Code of the same line from the second file, for all the lines in the documents.\n",
    "\n",
    "If both documents does not have the same number of lines to start, it is a bad sign, so we can check it first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File lenght does match!\n"
     ]
    }
   ],
   "source": [
    "with open('Germplasm.tsv') as germplasm_file, open('LocusGene.tsv') as locus_gene_file:\n",
    "    germplasm_lines = germplasm_file.readlines()\n",
    "    locus_gene_lines = locus_gene_file.readlines()\n",
    "    \n",
    "    if len(germplasm_lines) != len(locus_gene_lines):\n",
    "        print(\"File length does not match\")\n",
    "    else:\n",
    "        print(\"File lenght does match!\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we try to check line by line:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File lenght does match!\n",
      "Every Locus Code is on the same line in both files\n"
     ]
    }
   ],
   "source": [
    "with open('Germplasm.tsv') as germplasm_file, open('LocusGene.tsv') as locus_gene_file:\n",
    "    germplasm_lines = germplasm_file.readlines()\n",
    "    locus_gene_lines = locus_gene_file.readlines()\n",
    "        \n",
    "    # removing headers!\n",
    "    del germplasm_lines[0]\n",
    "    del locus_gene_lines[0]\n",
    "    \n",
    "    if len(germplasm_lines) != len(locus_gene_lines):\n",
    "        print(\"File length does not match\")\n",
    "    else:\n",
    "        print(\"File lenght does match!\")\n",
    "    \n",
    "    all_rows_match = True\n",
    "    \n",
    "    for line in range(len(germplasm_lines)):\n",
    "        code_germplasm = germplasm_lines[line].split(\"\\t\")[0]\n",
    "        code_locus_gene = locus_gene_lines[line].split(\"\\t\")[0]\n",
    "        \n",
    "        if code_germplasm != code_locus_gene:\n",
    "            all_rows_match = False\n",
    "            \n",
    "    if all_rows_match:\n",
    "        print(\"Every Locus Code is on the same line in both files\")\n",
    "    else:\n",
    "        print(\"There is some Locus Code in different lines\")\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2\n",
    "\n",
    "I will use sqlMagic to create the tables, so I need to run the following commands:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sql extension is already loaded. To reload it, use:\n",
      "  %reload_ext sql\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first step is creating the database and selecting it. I am going to name it \"exam_2\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql create database exam_2\n",
    "%sql use exam_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now it is time to create the tables of the database. I have checked the content of each field throught commands in a terminal. \n",
    "\n",
    "Nonetheless, it is interesting to know the maximum length of the phenotype description in order to create the table properly. I will do this next:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Max length for phenotype:  339\n",
      "Max length for germplasm name:  24\n"
     ]
    }
   ],
   "source": [
    "with open('Germplasm.tsv') as germplasm_file:\n",
    "    \n",
    "    phenotype_len = 0\n",
    "    germplasm_len = 0\n",
    "    \n",
    "    for line in germplasm_file.readlines():\n",
    "\n",
    "        if len(line.split(\"\\t\")[2]) > phenotype_len:\n",
    "            phenotype_len = len(line.split(\"\\t\")[2])\n",
    "            \n",
    "        if len(line.split(\"\\t\")[1]) > germplasm_len:\n",
    "            germplasm_len = len(line.split(\"\\t\")[1])\n",
    "            \n",
    "    print(\"Max length for phenotype: \", max_length)\n",
    "    print(\"Max length for germplasm name: \", germplasm_len)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Regarding the maximum lengths, I will set the number of varchar to 500 and 50 for the phenotype and germplasm name respectively.\n",
    "\n",
    "Trying to follow the 'DRY' rule, I will not insert the AGI Locus Code in both tables. Instead, I will reference from one of the tables the Locus Code that is stored in the other."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, locus_id INTEGER NOT NULL, germplasm VARCHAR(50) NOT NULL, phenotype VARCHAR(500) NOT NULL, pubmed INTEGER NOT NULL);\n",
    "%sql CREATE TABLE locus(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, locus VARCHAR(9) NOT NULL, gene VARCHAR(20) NOT NULL, protein_length VARCHAR(20) NOT NULL);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td>auto_increment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus_id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "        <td>varchar(50)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>phenotype</td>\n",
       "        <td>varchar(500)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>pubmed</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('id', 'int(11)', 'NO', 'PRI', None, 'auto_increment'),\n",
       " ('locus_id', 'int(11)', 'NO', '', None, ''),\n",
       " ('germplasm', 'varchar(50)', 'NO', '', None, ''),\n",
       " ('phenotype', 'varchar(500)', 'NO', '', None, ''),\n",
       " ('pubmed', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql describe germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td>auto_increment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus</td>\n",
       "        <td>varchar(9)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>gene</td>\n",
       "        <td>varchar(20)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>protein_length</td>\n",
       "        <td>varchar(20)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('id', 'int(11)', 'NO', 'PRI', None, 'auto_increment'),\n",
       " ('locus', 'varchar(9)', 'NO', '', None, ''),\n",
       " ('gene', 'varchar(20)', 'NO', '', None, ''),\n",
       " ('protein_length', 'varchar(20)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql describe locus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3\n",
    "\n",
    "I will use the readlines() method to read the .tsv files one more time, as I need to access the same number of line in both files and I do not know how to accomplish that with the csv library (as dictionary entries are not numbered).\n",
    "\n",
    "For further explanations of the code, see the comments embedded in it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_2',\n",
    "                             charset='utf8mb4',\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(False)\n",
    "\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor, open('Germplasm.tsv') as germplasm_file, open('LocusGene.tsv') as locus_file:\n",
    "        \n",
    "        germplasm_lines = germplasm_file.readlines()\n",
    "        locus_lines = locus_file.readlines()\n",
    "\n",
    "        # removing headers!\n",
    "        del germplasm_lines[0]\n",
    "        del locus_lines[0]\n",
    "\n",
    "        for line in range(len(germplasm_lines)):\n",
    "            # extracting data for table 'locus'\n",
    "            locus_code = locus_lines[line].split(\"\\t\")[0] # ATxGxxxxxx\n",
    "            locus_gene = locus_lines[line].split(\"\\t\")[1] # e.g. CIPK9\n",
    "            locus_protein_length = locus_lines[line].split(\"\\t\")[2].strip() # e.g. 223\n",
    "            # extracting data for table 'germplasm'\n",
    "            germplasm_name = germplasm_lines[line].split(\"\\t\")[1] # e.g. CS3828\n",
    "            germplasm_phenotype = germplasm_lines[line].split(\"\\t\")[2] # textual description\n",
    "            germplasm_pubmed = germplasm_lines[line].split(\"\\t\")[3].strip() # numerical code\n",
    "            \n",
    "            \n",
    "            # first, insert a line into 'locus' and retrieve its id\n",
    "            sql = \"\"\"INSERT INTO locus (locus, gene, protein_length) \n",
    "            VALUES ('{}', '{}', {});\"\"\".format(locus_code, locus_gene, locus_protein_length)\n",
    "            cursor.execute(sql)\n",
    "            sql = \"SELECT last_insert_id();\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            locus_id = results[0]['last_insert_id()']\n",
    "            \n",
    "            # then, insert the line in germplasm referencing the previous locus\n",
    "            sql = \"\"\"INSERT INTO germplasm (locus_id, germplasm, phenotype, pubmed) \n",
    "            VALUES ({}, '{}', '{}', {});\"\"\".format(locus_id, germplasm_name, germplasm_phenotype, germplasm_pubmed)\n",
    "            cursor.execute(sql)\n",
    "        \n",
    "        # commit once the loops ends\n",
    "        connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\\nInsertion is finished\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In case something goes wrong, you can delete the tables with the following command and run again Problem 2 to restore them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_exam_2</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql drop table germplasm\n",
    "%sql drop table locus\n",
    "%sql show tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check the inserted data. You can switch from 'germplasm' to 'locus' to check both tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>id</th>\n",
       "        <th>locus_id</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "        <td>1</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "        <td>2</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>3</td>\n",
       "        <td>3</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>4</td>\n",
       "        <td>4</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>5</td>\n",
       "        <td>5</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>6</td>\n",
       "        <td>6</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>7</td>\n",
       "        <td>7</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>8</td>\n",
       "        <td>8</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>9</td>\n",
       "        <td>9</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>10</td>\n",
       "        <td>10</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>11</td>\n",
       "        <td>11</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>12</td>\n",
       "        <td>12</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>13</td>\n",
       "        <td>13</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>14</td>\n",
       "        <td>14</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>15</td>\n",
       "        <td>15</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>16</td>\n",
       "        <td>16</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>17</td>\n",
       "        <td>17</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>18</td>\n",
       "        <td>18</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>19</td>\n",
       "        <td>19</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>20</td>\n",
       "        <td>20</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>21</td>\n",
       "        <td>21</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>22</td>\n",
       "        <td>22</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>23</td>\n",
       "        <td>23</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>24</td>\n",
       "        <td>24</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>25</td>\n",
       "        <td>25</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>26</td>\n",
       "        <td>26</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>27</td>\n",
       "        <td>27</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>28</td>\n",
       "        <td>28</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>29</td>\n",
       "        <td>29</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>30</td>\n",
       "        <td>30</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>31</td>\n",
       "        <td>31</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>32</td>\n",
       "        <td>32</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(1, 1, 'CS3828', 'Increased abundance of miRNA precursors.', 17369351),\n",
       " (2, 2, 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 16891401),\n",
       " (3, 3, 'SALK_058629', 'hypersensitive to low potassium media', 17486125),\n",
       " (4, 4, 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 18199744),\n",
       " (5, 5, 'SALK_042433', 'Multiple straight hairs', 16367956),\n",
       " (6, 6, 'gek1-1', 'Ethanol hypersensitivity.', 15215505),\n",
       " (7, 7, 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 17933900),\n",
       " (8, 8, 'pot1-1', 'No visible phenotype.', 17627276),\n",
       " (9, 9, 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 17419837),\n",
       " (10, 10, 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', 18484180),\n",
       " (11, 11, 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 21478444),\n",
       " (12, 12, 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 17586653),\n",
       " (13, 13, 'sep2-1', 'Non-described subtle phenotype.', 10821278),\n",
       " (14, 14, 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 17672843),\n",
       " (15, 15, 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770),\n",
       " (16, 16, 'vtc4-1', 'ascorbate deficient', 16595667),\n",
       " (17, 17, 'hdg8-1', 'No visible phenotype.', 16778018),\n",
       " (18, 18, 'pdd17', 'Defective pollen development.', 19237690),\n",
       " (19, 19, 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 16297066),\n",
       " (20, 20, 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 10982425),\n",
       " (21, 21, 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186),\n",
       " (22, 22, 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', 20388662),\n",
       " (23, 23, 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 17513499),\n",
       " (24, 24, 'prz1-1', 'Altered response to auxin and cytokinin', 12747832),\n",
       " (25, 25, 'atint4-2', 'No visible phenotype.', 16603666),\n",
       " (26, 26, 'pas2-3', 'Segregates 25% embryo lethal.', 18799749),\n",
       " (27, 27, 'plt3-1', 'Short roots and shortened root meristem.', 17960244),\n",
       " (28, 28, 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 18779216),\n",
       " (29, 29, 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 16891401),\n",
       " (30, 30, 'myb3r4-1', 'No visible phenotype.', 17287251),\n",
       " (31, 31, 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 19413687),\n",
       " (32, 32, 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 12345678)]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM germplasm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4\n",
    "\n",
    "I will format my file as \"Report number, blank line, solution\" for each of the asked reports.\n",
    "\n",
    "I will not close connection to MySQL server until I have finished all the requests.\n",
    "\n",
    "The values in the solution will be separated by commas, but due to the previously mentioned format I will not give the extention '.csv' to the file. It will be a '.txt' instead."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "Report 1\n",
      "\n",
      "id,locus_id,germplasm,phenotype,pubmed,locus.id,locus,gene,protein_length\n",
      "1,1,CS3828,Increased abundance of miRNA precursors.,17369351,1,AT1G01040,DCL1,332\n",
      "2,2,lhy-101,The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.,16891401,2,AT1G01060,LHY,290\n",
      "3,3,SALK_058629,hypersensitive to low potassium media,17486125,3,AT1G01140,CIPK9,223\n",
      "4,4,SALK_012400C,fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.,18199744,4,AT1G01220,FKGP,190\n",
      "5,5,SALK_042433,Multiple straight hairs,16367956,5,AT2G03720,MRH6,189\n",
      "6,6,gek1-1,Ethanol hypersensitivity.,15215505,6,AT2G03800,GEK1,196\n",
      "7,7,xerico,Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.,17933900,7,AT2G04240,XERICO,256\n",
      "8,8,pot1-1,No visible phenotype.,17627276,8,AT2G05210,POT1A,221\n",
      "9,9,rpk2-2,The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.,17419837,9,AT3G02130,RPK2,284\n",
      "10,10,afp4-1,Decreased germination on high concentrations of glucose and sorbitol.,18484180,10,AT3G02140,TMAC2,300\n",
      "11,11,rgp1-1,rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.,21478444,11,AT3G02230,RGP1,301\n",
      "12,12,tir3-1 RGLG1:rglg1 rglg2,The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.,17586653,12,AT3G02260,BIG,279\n",
      "13,13,sep2-1,Non-described subtle phenotype.,10821278,13,AT3G02310,SEP2,175\n",
      "14,14,atnbs1-1,Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.,17672843,14,AT3G02680,NBS1,190\n",
      "15,15,CS3816,The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.,17568770,15,AT3G02850,SKOR,234\n",
      "16,16,vtc4-1,ascorbate deficient,16595667,16,AT3G02870,VTC4,311\n",
      "17,17,hdg8-1,No visible phenotype.,16778018,17,AT3G03260,HDG8,194\n",
      "18,18,pdd17,Defective pollen development.,19237690,18,AT4G14790,SUV3,312\n",
      "19,19,bmy1-2,Plants cold-shocked for 6h have an increased starch content compared to wildtype.,16297066,19,AT4G15210,BAM5,313\n",
      "20,20,cla1-1,Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.,10982425,20,AT4G15560,DXS,219\n",
      "21,21,maa3,Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.,18772186,21,AT4G15570,MAA3,294\n",
      "22,22,Athspb-2,Early flowering, reduced fertility, aborted seeds.,20388662,22,AT4G15802,HSBP,254\n",
      "23,23,esd4-2,Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.,17513499,23,AT4G15880,ESD4,265\n",
      "24,24,prz1-1,Altered response to auxin and cytokinin,12747832,24,AT4G16420,ADA2B,279\n",
      "25,25,atint4-2,No visible phenotype.,16603666,25,AT4G16480,INT4,284\n",
      "26,26,pas2-3,Segregates 25% embryo lethal.,18799749,26,AT5G10480,PAS2,301\n",
      "27,27,plt3-1,Short roots and shortened root meristem.,17960244,27,AT5G10510,AIL6,310\n",
      "28,28,kns2,Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.,18779216,28,AT5G11110,SPS2,232\n",
      "29,29,hy5-101,Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.,16891401,29,AT5G11260,HY5,221\n",
      "30,30,myb3r4-1,No visible phenotype.,17287251,30,AT5G11510,MYB3R-4,336\n",
      "31,31,pyd2-2,The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.,19413687,31,AT5G12200,PYD2,310\n",
      "32,32,crn-1,Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.,12345678,32,AT5G13290,CRN,189\n",
      "\n",
      "Report 2\n",
      "\n",
      "15,15,CS3816,The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.,17568770,15,AT3G02850,SKOR,234\n",
      "21,21,maa3,Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.,18772186,21,AT4G15570,MAA3,294\n",
      "\n",
      "Report 3\n",
      "\n",
      "Number of matches for chromosome 1 : 4\n",
      "Number of matches for chromosome 2 : 4\n",
      "Number of matches for chromosome 3 : 9\n",
      "Number of matches for chromosome 4 : 8\n",
      "Number of matches for chromosome 5 : 7\n",
      "\n",
      "Report 4\n",
      "\n",
      "Average protein length for chromosome 1 : 258.75\n",
      "Average protein length for chromosome 2 : 215.5\n",
      "Average protein length for chromosome 3 : 252.0\n",
      "Average protein length for chromosome 4 : 277.5\n",
      "Average protein length for chromosome 5 : 271.2857142857143\n",
      "\n"
     ]
    }
   ],
   "source": [
    "separator = \",\"\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam_2',\n",
    "                             charset='utf8mb4',\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(False)\n",
    "\n",
    "# Report 1\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"SELECT * FROM germplasm FULL JOIN locus ON locus.id=locus_id\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        ### print(results)\n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "with open('Exam_2_reports.txt', 'w') as result_file:\n",
    "    result_file.write(\"Report 1\\n\\n\")\n",
    "    result_file.write(separator.join(results[0].keys())+\"\\n\")\n",
    "    \n",
    "    for line in results:\n",
    "        values = []\n",
    "        for key in line.keys():\n",
    "            values.append(str(line[key]))\n",
    "        result_file.write(separator.join(values)+\"\\n\")\n",
    "        \n",
    "# Report 2\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM germplasm FULL JOIN locus ON locus.id=locus_id\n",
    "         WHERE gene='MAA3' OR gene='SKOR';\"\"\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        ### print(results)\n",
    "finally:\n",
    "    print(\"\")\n",
    "    \n",
    "with open('Exam_2_reports.txt', 'a') as result_file:\n",
    "    result_file.write(\"\\nReport 2\\n\\n\")\n",
    "    \n",
    "    for line in results:\n",
    "        values = []\n",
    "        for key in line.keys():\n",
    "            values.append(str(line[key]))\n",
    "        result_file.write(separator.join(values)+\"\\n\")\n",
    "        \n",
    "# Report 3\n",
    "try:\n",
    "    with connection.cursor() as cursor, open('Exam_2_reports.txt', 'a') as result_file:\n",
    "        \n",
    "        result_file.write(\"\\nReport 3\\n\\n\")\n",
    "        \n",
    "        for number in ['1', '2', '3', '4', '5']:\n",
    "            sql = \"\"\"SELECT COUNT(*) AS \"Number of matches for chromosome {}\" FROM locus\n",
    "             WHERE locus.locus LIKE 'AT{}G%'\"\"\".format(number, number)\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            ### print(results)\n",
    "            \n",
    "            for (key, value) in results[0].items():\n",
    "                result_file.write(\"{} : {}\\n\".format(key, value))\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "\n",
    "# Report 4\n",
    "try:\n",
    "    with connection.cursor() as cursor, open('Exam_2_reports.txt', 'a') as result_file:\n",
    "        \n",
    "        result_file.write(\"\\nReport 4\\n\\n\")\n",
    "        \n",
    "        for number in ['1', '2', '3', '4', '5']:\n",
    "            sql = \"\"\"SELECT AVG(locus.protein_length) AS \"Average protein length for chromosome {}\" FROM locus\n",
    "             WHERE locus.locus LIKE 'AT{}G%'\"\"\".format(number, number)\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            ### print(results)\n",
    "            \n",
    "            for (key, value) in results[0].items():\n",
    "                result_file.write(\"{} : {}\\n\".format(key, value))\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "        \n",
    "        \n",
    "connection.close()\n",
    "        \n",
    "with open('Exam_2_reports.txt', 'r') as result_file:\n",
    "    print(result_file.read())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
